File: script_parser.py
from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum


class ActionKind(str, Enum):
    TEXT = "text"
    KEY = "key"
    HOTKEY = "hotkey"
    WAIT = "wait"


@dataclass(frozen=True)
class Action:
    kind: ActionKind
    value: str = ""
    duration_ms: int = 0


KEY_NAMES = {
    "BACKSPACE": "backspace",
    "DELETE": "delete",
    "DOWN": "down",
    "END": "end",
    "ENTER": "enter",
    "ESC": "esc",
    "ESCAPE": "esc",
    "HOME": "home",
    "LEFT": "left",
    "PAGEDOWN": "pagedown",
    "PAGEUP": "pageup",
    "RIGHT": "right",
    "SPACE": "space",
    "TAB": "tab",
    "UP": "up",
    **{f"F{number}": f"f{number}" for number in range(1, 25)},
}
MODIFIERS = {"ALT", "CTRL", "CONTROL", "SHIFT", "WIN", "WINDOWS"}
WAIT_PATTERN = re.compile(r"^WAIT(?:\s+|:)(\d{1,7})$", re.IGNORECASE)


def _append_text(actions: list[Action], text: str) -> None:
    if not text:
        return
    if actions and actions[-1].kind is ActionKind.TEXT:
        previous = actions[-1]
        actions[-1] = Action(ActionKind.TEXT, previous.value + text)
    else:
        actions.append(Action(ActionKind.TEXT, text))


def _parse_token(token: str) -> Action | None:
    normalized = token.strip().upper()
    if normalized in KEY_NAMES:
        return Action(ActionKind.KEY, KEY_NAMES[normalized])

    wait_match = WAIT_PATTERN.fullmatch(normalized)
    if wait_match:
        return Action(ActionKind.WAIT, duration_ms=int(wait_match.group(1)))

    parts = [part.strip().upper() for part in normalized.split("+")]
    if len(parts) >= 2 and all(parts) and all(part in MODIFIERS for part in parts[:-1]):
        aliases = {
            "CONTROL": "ctrl",
            "CTRL": "ctrl",
            "WINDOWS": "windows",
            "WIN": "windows",
            "ALT": "alt",
            "SHIFT": "shift",
        }
        final = KEY_NAMES.get(parts[-1], parts[-1].lower())
        return Action(
            ActionKind.HOTKEY,
            "+".join([aliases[part] for part in parts[:-1]] + [final]),
        )
    return None


def parse_script(script: str) -> list[Action]:
    """Parse AutoTyper tokens while leaving unknown braces as ordinary text."""
    actions: list[Action] = []
    index = 0
    while index < len(script):
        if script.startswith("{{", index):
            _append_text(actions, "{")
            index += 2
            continue
        if script.startswith("}}", index):
            _append_text(actions, "}")
            index += 2
            continue
        if script[index] != "{":
            next_open = script.find("{", index)
            next_close = script.find("}}", index)
            candidates = [position for position in (next_open, next_close) if position >= 0]
            end = min(candidates) if candidates else len(script)
            _append_text(actions, script[index:end])
            index = end
            continue

        close = script.find("}", index + 1)
        if close < 0:
            _append_text(actions, script[index:])
            break
        token_text = script[index + 1 : close]
        action = _parse_token(token_text)
        if action is None:
            _append_text(actions, script[index : close + 1])
        else:
            actions.append(action)
        index = close + 1
    return actions


def action_units(actions: list[Action]) -> int:
    return sum(
        len(action.value) if action.kind is ActionKind.TEXT else 1
        for action in actions
        if action.kind is not ActionKind.WAIT
    )


def describe_tokens() -> str:
    return (
        "Tokens: {ENTER}  {TAB}  {WAIT 500}  {CTRL+ENTER}  "
        "{UP}/{DOWN}. Use {{ for a literal brace."
    )

File: web/client/src/lib/scriptParser.js
/**
 * Parsing utilities shared by the editor preview and the playback engine.
 *
 * A parsed script is intentionally a flat list of units.  That makes timing,
 * progress reporting, pause/resume, and target-window mutations use exactly
 * the same representation.  Text units are grapheme clusters rather than
 * UTF-16 code units, so an emoji or a combining-character sequence is never
 * split across two ticks.
 */

const SPECIAL_KEY_TOKENS = new Map([
  ["tab", "tab"],
  ["enter", "enter"],
  ["return", "enter"],
  ["space", "space"],
  ["spacebar", "space"],
  ["backspace", "backspace"],
  ["bs", "backspace"],
  ["bksp", "backspace"],
  ["delete", "delete"],
  ["del", "delete"],
  ["left", "arrowleft"],
  ["arrowleft", "arrowleft"],
  ["right", "arrowright"],
  ["arrowright", "arrowright"],
  ["up", "arrowup"],
  ["arrowup", "arrowup"],
  ["down", "arrowdown"],
  ["arrowdown", "arrowdown"],
  ["home", "home"],
  ["end", "end"],
  ["escape", "escape"],
  ["esc", "escape"],
]);

const MAX_WAIT_MS = 86_400_000; // One day; safely below an unbounded input.

let graphemeSegmenter;

/**
 * Return user-perceived characters.  Modern browsers and supported Node
 * versions expose Intl.Segmenter; the fallback still keeps surrogate pairs
 * together when it is unavailable.
 */
export function segmentGraphemes(value) {
  const text = value == null ? "" : String(value);
  if (!text) {
    return [];
  }

  if (typeof Intl !== "undefined" && typeof Intl.Segmenter === "function") {
    graphemeSegmenter ??= new Intl.Segmenter(undefined, {
      granularity: "grapheme",
    });
    return Array.from(graphemeSegmenter.segment(text), (part) => part.segment);
  }

  return Array.from(text);
}

function parseToken(rawToken) {
  const normalized = rawToken.trim().toLowerCase();
  const key = SPECIAL_KEY_TOKENS.get(normalized);
  if (key) {
    return { kind: "key", value: key };
  }

  // WAIT is not required for the assignment, but costs no extra dependency
  // and mirrors the desktop predecessor's useful script syntax. Unknown or
  // out-of-range wait-like tokens intentionally remain literal text.
  const wait = /^wait(?:\s+|:)\s*(\d{1,9})$/i.exec(rawToken.trim());
  if (wait) {
    const durationMs = Number(wait[1]);
    if (Number.isSafeInteger(durationMs) && durationMs <= MAX_WAIT_MS) {
      return { kind: "wait", durationMs };
    }
  }

  return null;
}

/**
 * Parse a script into playback units.
 *
 * Supported tokens are case-insensitive: {Tab}, {Enter}, {Space}, plus a few
 * harmless editing/navigation keys. {{ and }} escape literal braces. Unknown
 * or unclosed tokens are emitted as ordinary text so templates such as
 * "Hello {customer_name}" are never corrupted.
 */
export function parseScript(script) {
  const source = script == null ? "" : String(script);
  const units = [];
  let bufferedText = "";

  const appendText = (text) => {
    bufferedText += text;
  };

  const flushText = () => {
    if (!bufferedText) {
      return;
    }
    for (const value of segmentGraphemes(bufferedText)) {
      units.push({ kind: "text", value });
    }
    bufferedText = "";
  };

  let index = 0;
  while (index < source.length) {
    if (source.startsWith("{{", index)) {
      appendText("{");
      index += 2;
      continue;
    }

    if (source.startsWith("}}", index)) {
      appendText("}");
      index += 2;
      continue;
    }

    if (source[index] !== "{") {
      appendText(source[index]);
      index += 1;
      continue;
    }

    const closeIndex = source.indexOf("}", index + 1);
    if (closeIndex < 0) {
      appendText(source.slice(index));
      break;
    }

    const rawToken = source.slice(index + 1, closeIndex);
    const unit = parseToken(rawToken);
    if (unit) {
      flushText();
      units.push(unit);
    } else {
      appendText(source.slice(index, closeIndex + 1));
    }
    index = closeIndex + 1;
  }

  flushText();
  return units;
}

/** Count output actions for progress bars; waits deliberately do not count. */
export function countTypingUnits(units) {
  if (!Array.isArray(units)) {
    return 0;
  }
  return units.reduce(
    (count, unit) =>
      count + (unit?.kind === "text" || unit?.kind === "key" ? 1 : 0),
    0,
  );
}

export function isTypingUnit(unit) {
  return unit?.kind === "text" || unit?.kind === "key";
}

export const supportedScriptTokens = Object.freeze([
  "{Tab}",
  "{Enter}",
  "{Space}",
  "{Backspace}",
  "{Delete}",
  "{Left}",
  "{Right}",
  "{Up}",
  "{Down}",
]);
